pkg/protocol/protocol_blocks.go

```diff
@@ -0,0 +1,153 @@
+				if replayBlocks {
+					for _, droppedBlock := range b.droppedBlocksBuffer.GetValues(commitment.ID()) {
+						b.LogTrace("replaying dropped block", "commitmentID", commitment.ID(), "blockID", droppedBlock.A.ID())
+
+						b.ProcessResponse(droppedBlock.A, droppedBlock.B)
+					}
+				}
+			})
+		})
+
+		protocol.Chains.WithInitializedEngines(func(chain *Chain, engine *engine.Engine) (shutdown func()) {
+			return lo.Batch(
+				engine.Events.BlockRequester.Tick.Hook(b.SendRequest).Unhook,
+				engine.Events.Scheduler.BlockScheduled.Hook(func(block *blocks.Block) { b.SendResponse(block.ModelBlock()) }).Unhook,
+				engine.Events.Scheduler.BlockSkipped.Hook(func(block *blocks.Block) { b.SendResponse(block.ModelBlock()) }).Unhook,
+			)
+		})
+	})
+
+	return b
+}
+
+// SendRequest sends a request for the given block to all peers.
+func (b *BlocksProtocol) SendRequest(blockID iotago.BlockID) {
+	b.workerPool.Submit(func() {
+		b.protocol.Network.RequestBlock(blockID)
+
+		b.LogTrace("request", "blockID", blockID)
+	})
+}
+
+// SendResponse sends the given block to all peers.
+func (b *BlocksProtocol) SendResponse(block *model.Block) {
+	b.workerPool.Submit(func() {
+		b.protocol.Network.SendBlock(block)
+
+		b.LogTrace("sent", "blockID", block.ID())
+	})
+}
+
+// ProcessResponse processes the given block response.
+func (b *BlocksProtocol) ProcessResponse(block *model.Block, from peer.ID) {
+	b.workerPool.Submit(func() {
+		// abort if the commitment belongs to an evicted slot
+		commitment, err := b.protocol.Commitments.Get(block.ProtocolBlock().Header.SlotCommitmentID, true)
+		if err != nil && ierrors.Is(ErrorSlotEvicted, err) {
+			b.LogError("dropped block referencing unsolidifiable commitment", "commitmentID", block.ProtocolBlock().Header.SlotCommitmentID, "blockID", block.ID(), "err", err)
+
+			return
+		}
+
+		// add the block to the dropped blocks buffer if we could not dispatch it to the chain
+		if commitment == nil || !commitment.Chain.Get().DispatchBlock(block, from) {
+			if !b.droppedBlocksBuffer.Add(block.ProtocolBlock().Header.SlotCommitmentID, types.NewTuple(block, from)) {
+				b.LogError("failed to add dropped block referencing unsolid commitment to dropped blocks buffer", "commitmentID", block.ProtocolBlock().Header.SlotCommitmentID, "blockID", block.ID())
+			} else {
+				b.LogTrace("dropped block referencing unsolid commitment added to dropped blocks buffer", "commitmentID", block.ProtocolBlock().Header.SlotCommitmentID, "blockID", block.ID())
+			}
+
+			return
+		}
+
+		b.LogTrace("received block", "blockID", block.ID(), "commitment", commitment.LogName())
+	})
+}
+
+// ProcessRequest processes the given block request.
+func (b *BlocksProtocol) ProcessRequest(blockID iotago.BlockID, from peer.ID) {
+	b.workerPool.Submit(func() {
+		block, exists := b.protocol.Engines.Main.Get().Block(blockID)
+
+		// TODO: CHECK why blocks can be nil here:
```

Comment:
Let's remove this, there's an issue #556 that tracks this already. I wouldn't even add the check for nil or at least explicitly log it.